Add CORE-V headers

PaoloS02 opened the pull request:
This is a reviewed version of the pull request #78

melonedo committed:
[RISCV] Add CORE-V headers

Add header files for all CORE-V extensions that have C builtin functions.
The definitions of existing builtin functions are also adjusted accordingly.

PaoloS02 committed:
[RISCV] Update the CORE-V C API headers to follow the RISC-V conventions.

This patch changes the C API headers that were added for the CORE-V ISA
extensions so that they follow the standard described for RISC-V here:

https://github.com/riscv-non-isa/riscv-c-api-doc/blob/master/riscv-c-api.md#intrinsic-functions

The C API headers must provide a uniform interface for instrinsic
functions that is compatible with any compiler that supports them.

In the wrapper of a C API intrinsic function definition any compiler
can implement the functionality differently, whether it is with an existing
instrinsic function, or with inline assembly etc...

The user code that adds a C API RISC-V header is agnostic to the
implementation and is then compatible with any compiler that has the header.

The RISC-V intrinsic functions have
- the common prefix '__riscv_' to avoid name collision with other targets.
- a vendor specific prefix if they belong to a vendor specific extension.
- the name of the function.
- the type of the function: this is typically used to distinguish
  different versions of the same intrinsic function that differ only for
  the type of one or more operands.

Another convention is to use the type 'long' instead of integer types
like int32_t or uint32_t as the type 'long' will be set to the
current XLEN.

According to these conventions the CORE-V intrinsic functions are
now redefined like:

long __riscv_cv_name_type(long op1, long op2) {
  return __builtin_riscv_cv_name_type(op1, op2);
}

lewis-revill commented on clang/lib/Headers/riscv_corev_alu.h:
```diff
@@ -0,0 +1,111 @@
+
+#if defined(__riscv_xcvalu)
+
+#define __DEFAULT_FN_ATTRS __attribute__((__always_inline__, __nodebug__))
```
This isn't a special symbol, so `__DEFAULT_FN_ATTRS` does also need to be added to each declaration rather than just defined here.

lewis-revill replied:
Each definition should also be `static inline`  (or, as appears to be the case for other targets' definitions, `static __inline__`)

PaoloS02 committed:
[RISCV] Use macros as C API interfaces of CORE-V intrinsics with const arguments.

This patch adds macros in the C API header files instead of wrapped
calls for those intrinsics which use const arguments.
This change affects the intrinsics for:

SIMD:

- add_h
- sub_h
- cplxmul_r
- cplxmul_i
- subrotmj
- shuffle_xhi
- extract[u]_[h/b]
- insert_[h/b]

Bitmanip:

- bitrev

It also changes the type of the immediate argument of extract and shuffle.sci
to const in the clang intrinsics definitions.

PaoloS02 committed:
[RISCV] Apply attributes to CORE-V C API intrinsic definitions.

PaoloS02 committed:
[RISCV] Remove inclusion of C API headers in CORE-V intrinsics tests.

This is done to separate the tests for the internal implementation
of the CORE-V intrinsics from the tests for C API interfaces of the
same.

ChunyuLiao commented:
I see some testcases failed,
<img width="589" alt="image" src="https://github.com/openhwgroup/corev-llvm-project/assets/26795311/a6ed06dd-4543-4352-99dc-7d5fe06a5de0">

PaoloS02 committed:
[RISCV] Add type casts and brackets in C API CORE-V simd and mac macros.

PaoloS02 committed:
[RISCV] Use macros in C API headers for mac intrinsics with constant arguments.

PaoloS02 committed:
[RISCV] Add CodeGen tests for the CORE-V C API headers.

PaoloS02 committed:
[RISCV] Set CORE-V intrinsic wrappers in C API headers as static.

This is done to avoid multiple defines in the link step if the inlining fails.

PaoloS02 committed:
[RISCV] Set CORE-V intrinsic wrappers in C API headers as __inline__.

PaoloS02 commented:
Implemented in new patches.
